exercise1/terraform/main.tf

```diff
@@ -0,0 +1,505 @@
+
+  # prevent the possibility of accidental data loss
+  lifecycle {
+    prevent_destroy = false # for the sake of testing
+  }
+}
+
+#_______________________ Redis Database resources _______________________________________________
+resource "azurerm_redis_cache" "ex1_vm_redis" {
+  name                          = "${var.rg_name}_redis"
+  resource_group_name           = azurerm_resource_group.ex1.name
+  location                      = azurerm_resource_group.ex1.location
+  capacity                      = 1
+  family                        = "C"
+  sku_name                      = "Basic"
+  public_network_access_enabled = false
+}
+
+# ************************** VM associated resources *******************************************
+
+#___________________________ VM network resources ______________________________
+resource "azurerm_subnet" "ex1_subnet_vm" {
+  name                 = "${var.rg_name}_subnet_vm"
+  resource_group_name  = azurerm_resource_group.ex1.name
+  virtual_network_name = azurerm_virtual_network.ex1_vnet.name
+  address_prefixes     = ["10.0.2.0/24"]
+}
+
+resource "azurerm_network_security_group" "ex1_vm_netsecg" {
+  name                = "${var.rg_name}_vm_netsecg"
+  location            = azurerm_resource_group.ex1.location
+  resource_group_name = azurerm_resource_group.ex1.name
+}
+
+resource "azurerm_subnet_network_security_group_association" "ex1_secg_asso_vm" {
+  subnet_id                 = azurerm_subnet.ex1_subnet_vm.id
+  network_security_group_id = azurerm_network_security_group.ex1_vm_netsecg.id
+}
+
+resource "azurerm_network_security_rule" "ex1_netsec_r_vm_443" {
+  name                        = "${var.rg_name}_netsec_r_443"
+  priority                    = 100
+  direction                   = "Inbound"
+  access                      = "Allow"
+  protocol                    = "Tcp"
+  source_port_range           = "443"
+  destination_port_range      = "443"
+  source_address_prefix       = "*"
+  destination_address_prefix  = "*"
+  resource_group_name         = azurerm_resource_group.ex1.name
+  network_security_group_name = azurerm_network_security_group.ex1_vm_netsecg.name
+}
+
+resource "azurerm_network_interface" "ex1_nic_vm" {
+  name                = "${var.rg_name}_nic_vm"
+  resource_group_name = azurerm_resource_group.ex1.name
+  location            = azurerm_resource_group.ex1.location
+
+  ip_configuration {
+    name                          = local.vm_nic_ip_name
+    subnet_id                     = azurerm_subnet.ex1_subnet_vm.id
+    private_ip_address_allocation = "Dynamic"
+  }
+}
+
+resource "azurerm_network_interface_security_group_association" "ex1_vm_netsecg_nic_asso" {
+  network_interface_id      = azurerm_network_interface.ex1_nic_vm.id
+  network_security_group_id = azurerm_network_security_group.ex1_vm_netsecg.id
+}
+
+#___________________________ VM associated resources ______________________________
+
+resource "azurerm_virtual_machine" "ex1_vm" {
```

Comment:
Typically all config values should be variables but for the purposes of this exercise, hard coded is fine.

Reply:
I prefer writing production quality code :)

I added vars for almost everything but ideally it would be one big object at the varaibales.tf level